### PR TITLE
[14.0][FIX] mrp_bom_attribute_match

### PR DESCRIPTION
**Pull request by ilyasProgrammer**
related to https://github.com/OCA/manufacture/issues/997
wip

New tests was copied from @felixhummel fork. @felixhummel do you mind? I will add you as a contributor.

### CODE REVIEW

**Comment by ilyasProgrammer** [comment]
@felixhummel Feel free to test ;)

**Comment by francesco-ooops** [comment]
> @felixhummel Feel free to test ;)

@lk-eska too :)

File: mrp_bom_attribute_match/reports/mrp_report_bom_structure.py [hunk 1]
```diff
@@ -43,3 +42,60 @@ def _get_bom_lines(self, bom, bom_quantity, product, line_id, level):
...
+
+    def _get_bom_lines(self, bom, bom_quantity, product, line_id, level):
+        components = []
+        total = 0
+        for line in bom.bom_line_ids:
+            line_quantity = (bom_quantity / (bom.product_qty or 1.0)) * line.product_qty
+            if line._skip_bom_line(product):
+                continue
+            company = bom.company_id or self.env.company
+            product_id = bom.get_component_template_product(
```
Comment (giarve) [inline mrp_bom_attribute_match/reports/mrp_report_bom_structure.py]: Shouldn't this function be `_get_component_template_product`? With a leading underscore.

Tests are passing, but this raises an error in Odoo

Comment (ilyasProgrammer) [reply mrp_bom_attribute_match/reports/mrp_report_bom_structure.py]: Could you share error text ?

Comment (giarve) [reply mrp_bom_attribute_match/reports/mrp_report_bom_structure.py]: > Could you share error text ?

https://github.com/OCA/manufacture/blob/b2c30a20aacda7f9fda6d964ac3f3471374403a7/mrp_bom_attribute_match/models/mrp_bom.py#L321

The function name does not exist (it is just different, with a leading underscore).

Comment (ilyasProgrammer) [reply mrp_bom_attribute_match/reports/mrp_report_bom_structure.py]: Ah yes right. Thanks!

**Comment by giarve** [comment]
For some reason I still do not see lines in the report (they are skipped) when I set a product component (product template), but that might be from my side. If someone else is willing to review if changes are working for them, I would appreciate.

**Comment by francesco-ooops** [comment]
@giarve did you test on runboat? can you post a video or a screenshot?

**Comment by giarve** [comment]
> @giarve did you test on runboat? can you post a video or a screenshot?

On the still running instance:

image

I believe it is because when adding variants the line where it appears is not removed (which is supposed to happen), but it still looks for a product_id and it is not found.

Also might have to do with dynamically created attributes? Although in this case they were instantly created. But I do not know if they are properly handled when product components are there but their dynamically created variants are not.

Something extra: in the structure & cost report, children with product components cannot be unfolded.


Summary: I am not sure this PR is the solution or it is just part of it.

@ilyasProgrammer I think the `def _get_bom_lines(self, bom, bom_quantity, product, line_id, level):` function is duplicated in the same file.

**Review by raistdev — APPROVED** [review]

**Comment by lk-eska** [comment]
I sent the fix here #997

**Comment by giarve** [comment]
> I sent the fix here #997

I am not sure whether it works for all use cases. It was the first fix that was proposed in this PR if I recall correctly.

File: mrp_bom_attribute_match/reports/mrp_report_bom_structure.py [hunk 2]
```diff
@@ -31,9 +31,7 @@ def _get_bom_lines(self, bom, bom_quantity, product, line_id, level):
             if to_ignore_line_ids:
                 for to_ignore_line_id in to_ignore_line_ids:
                     bom.bom_line_ids = [(3, to_ignore_line_id, 0)]
-        product = bom.get_component_template_product(
-            line, product, line.product_id
-        )
+        product = bom.get_component_template_product(line, product, line.product_id)
```
Comment (giarve) [inline mrp_bom_attribute_match/reports/mrp_report_bom_structure.py]: Please check the function name, shouldn't it start with the underscore?

Maybe there is something I am missing here

**Comment by ilyasProgrammer** [comment]
@giarve Please try again

**Comment by lk-eska** [comment]
> > I sent the fix here #997
> 
> I am not sure whether it works for all use cases. It was the first fix that was proposed in this PR if I recall correctly.

no it was something else... anyway this is coming to an end. if it works we would be happy to use it.

**Review by giarve — APPROVED** [review]
Works beautifully from my side. I think we could already merge this and if in the future something else pops up we would rather create another PR (hopefully not needed!).

**Comment by francesco-ooops** [comment]
@ivantodorovich good for merge?

File: mrp_bom_attribute_match/reports/mrp_report_bom_structure.py [hunk 3]
```diff
@@ -43,3 +46,67 @@ def _get_bom_lines(self, bom, bom_quantity, product, line_id, level):
                     if isinstance(value, models.NewId):
                         component[key] = value.origin
         return components, total
+
+    def _get_price(self, bom, factor, product):
```
Comment (ivantodorovich) [inline mrp_bom_attribute_match/reports/mrp_report_bom_structure.py:50]: Would it be possible to add a few inline comments to indicate the patched lines?
As this is another complete override, it's always useful to know exactly what changed, specially for migrations.

**Review by ivantodorovich — APPROVED** [review]
LG ! 
Thanks @ilyasProgrammer  :)

File: mrp_bom_attribute_match/reports/mrp_report_bom_structure.py (hunk 3 above)
Comment (ilyasProgrammer) [reply mrp_bom_attribute_match/reports/mrp_report_bom_structure.py:50]: Agree

**Review by ivantodorovich — APPROVED** [review]

**Comment by francesco-ooops** [comment]
@ilyasProgrammer can you try again?

**Pull request merged**